main.js

```diff
@@ -0,0 +1,271 @@
+/* CONSTANTS AND GLOBALS */
+const canvas = document.getElementById("canvas"),
+  context = canvas.getContext("2d"),
+  controlsIconWidth = 30,
+  controlsWidth = controlsIconWidth * 3,
+  controlPanelWidth = 200,
+  width = (canvas.width = window.innerWidth),
+  height = (canvas.height = window.innerHeight),
+  opacityAlpha = 0.0075,
+  center = [width / 2, height / 2],
+  friction = 0.99,
+  scale = 0.003,
+  ease = d3.easeCubic;
+
+let controlContainer;
+let animationTimer = false;
+let res = 10
+let points = [];
+
+/* INITIAL STATE */
+let state = {
+  controlsOpen: false,
+  numPoints: 1200,
+  speed: 0.025,
+  showFlow: true,
+  paused: false,
+};
+
+// user controls
+const controls = {
+  numPoints: {
+    type: "text",
+    display: "number of lines",
+  },
+  showFlow: {
+    type: "checkbox",
+    display: "show flow fields"
+  }
+};
+
+// run the thang
+init()
+
+/* SET STATE FUNCTION */
+function setStateAndDraw(nextState) {
+  state = Object.assign({}, state, nextState);
+  console.log("state update", state);
+  draw();
+}
+
+function setStateOnly(nextState, callback = null) {
+  state = Object.assign({}, state, nextState);
+  console.log("state update", state);
+  callback;
```

Comment:
safer to pass name of function, then invoke it here, instead of invoking it when you call `setStateOnly`, because you want the callback invoked *after* state changes

Reply:
nice good call